kubeflow/trainer/api/trainer_client.py

```diff
@@ -12,526 +12,137 @@
+from typing import Dict, Optional, Union, Set, List
 
 from kubeflow.trainer.constants import constants
 from kubeflow.trainer.types import types
-from kubeflow.trainer.utils import utils
-from kubeflow_trainer_api import models
-from kubernetes import client, config, watch
+from kubeflow.trainer.types.backends import K8SBackendConfig, LocalProcessBackendConfig
+from kubeflow.trainer.backends import TRAINER_BACKENDS
+
 
 logger = logging.getLogger(__name__)
 
+BackendCfg = Union[K8SBackendConfig, LocalProcessBackendConfig]
+
 
 class TrainerClient:
-    def __init__(
-        self,
-        config_file: Optional[str] = None,
-        context: Optional[str] = None,
-        client_configuration: Optional[client.Configuration] = None,
-        namespace: Optional[str] = None,
+
+    def __init__( self,
+            backend_config: Optional[BackendCfg] = K8SBackendConfig()
     ):
-        """TrainerClient constructor. Configure logging in your application
-            as follows to see detailed information from the TrainerClient APIs:
-            .. code-block:: python
-                import logging
-                logging.basicConfig()
-                log = logging.getLogger("kubeflow.trainer.api.trainer_client")
-                log.setLevel(logging.DEBUG)
+        """
+        Initialize a trainer client.
 
         Args:
-            config_file: Path to the kube-config file. Defaults to ~/.kube/config.
-            context: Set the active context. Defaults to current_context from the kube-config.
-            client_configuration: Client configuration for cluster authentication.
-                You have to provide valid configuration with Bearer token or
-                with username and password. You can find an example here:
-                https://github.com/kubernetes-client/python/blob/67f9c7a97081b4526470cad53576bc3b71fa6fcc/examples/remote_cluster.py#L31
-            namespace: Target Kubernetes namespace. If SDK runs outside of Kubernetes cluster it
-                takes the namespace from the kube-config context. If SDK runs inside
-                the Kubernetes cluster it takes namespace from the
-                `/var/run/secrets/kubernetes.io/serviceaccount/namespace` file. By default it
-                uses the `default` namespace.
+            backend_config: Backend configuration. Either K8SBackendConfig or
+                            LocalProcessBackendConfig, or None to use the backend's
+                            default config class. Defaults to K8SBackendConfig.
         """
+        # initialize training backend
+        self.__backend = self.__init_backend(backend_config)
 
-        if namespace is None:
-            namespace = utils.get_default_target_namespace(context)
+    def __init_backend(self, backend_config: BackendCfg):
+        backend = TRAINER_BACKENDS.get(backend_config.__class__)
+        if not backend:
+            raise ValueError("Invalid backend config '{}'".format(backend_config))
```

Comment:
Do you need to keep this nested function to initializer backend ?
```suggestion
        backend = TRAINER_BACKENDS.get(backend_config.__class__)
        if not backend:
            raise ValueError("Invalid backend config '{}'".format(backend_config))
```

Reply:
not sure what is needed here